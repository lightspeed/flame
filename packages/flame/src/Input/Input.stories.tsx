import * as React from 'react';

import { Input } from './Input';

import { Box } from '../Core';
import { Badge } from '../Badge';
import { Button } from '../Button';
import { Icon } from '../Icon';

export default {
  title: 'Components/Input',
  component: Input,
  argTypes: {
    onChange: { action: 'on change' },
    onFocus: { action: 'on focus' },
    onBlur: { action: 'on blur' },
  },
};

const InputWrapper = () => {
  const [state, setState] = React.useState<{
    status: {
      type?: 'error' | 'valid' | 'warning';
      message?: string;
    };
    input: any;
  }>({
    status: {},
    input: '',
  });

  const handleClick = () => {
    if (
      !state.input ||
      (typeof state.input === 'string' && Number.isNaN(parseFloat(state.input)))
    ) {
      setState(prevState => ({
        ...prevState,
        status: { type: 'error', message: 'Please enter a valid number' },
      }));
    } else if (state.input < 10) {
      setState(prevState => ({
        ...prevState,
        status: { type: 'warning', message: 'You can do more than 10!' },
      }));
    } else {
      setState(prevState => ({ ...prevState, status: { type: 'valid', message: 'Good job!' } }));
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const target = e.target;
    setState(prevState => ({ ...prevState, input: target.value }));
  };

  const handleFocus = () => {
    setState(prevState => ({ ...prevState, status: {} }));
  };

  return (
    <div>
      <Input
        placeholder="Input with error message..."
        id="error"
        onChange={handleChange}
        onFocus={handleFocus}
        label="Enter a number"
        status={state.status.type as any}
        statusMessage={state.status.message}
      />
      <br />
      <Button onClick={() => handleClick()}>Submit</Button>
    </div>
  );
};

export const story = () => (
  <div>
    <Box mb={3}>
      <h3>Simple text input</h3>
      <Input id="simple" label="Label" placeholder="Placeholder text..." />
    </Box>
    <Box mb={3}>
      <h3>Advanced text input</h3>
      <Input
        id="advanced"
        label="Label"
        labelHelper={
          <Badge type="info" size="small">
            Label Helper
          </Badge>
        }
        textHelper="Text Helper"
        description="Description text"
        placeholder="Placeholder text..."
        prefix="$"
        suffix="*"
      />
    </Box>

    <h2>States</h2>
    <Box mb={3}>
      <Input placeholder="Placeholder text..." />
    </Box>
    <Box mb={3}>
      <Input placeholder="Hover input..." />
    </Box>
    <Box mb={3}>
      <Input placeholder="Focus input..." />
    </Box>
    <Box mb={3}>
      <Input readOnly value="Readonly input" />
    </Box>
    <Box mb={3}>
      <Input disabled placeholder="Disabled input..." />
    </Box>
    <Box mb={3}>
      <Input
        status="valid"
        placeholder="Valid input..."
        prefix={<Icon name="customers" />}
        suffix={<Icon name="info" color="secondary" />}
      />
    </Box>
    <Box mb={3}>
      <Input status="warning" placeholder="Warning input..." />
    </Box>
    <Box mb={3}>
      <Input status="error" placeholder="Error input..." />
    </Box>

    <div>
      <h2>Size</h2>
      <Box mb={3}>
        <Input size="small" placeholder="Small input..." />
      </Box>
      <Box mb={3}>
        <Input placeholder="Normal input..." />
      </Box>
      <Box mb={3}>
        <Input size="large" placeholder="Large input..." />
      </Box>
    </div>
    <div>
      <h2>Input Error Text</h2>
      <Box mb={3}>
        <Input
          placeholder="This is some text on an input that has an error"
          status="error"
          statusMessage="This is an error message"
        />
      </Box>
    </div>
    <div>
      <h2>Input Error Text (using the legacy API)</h2>
      <Box mb={3}>
        <Input
          placeholder="This is some text on an input that has an error"
          status={{ type: 'error', message: 'This is an error message' }}
        />
      </Box>
    </div>
  </div>
);

export const events: React.FC<{
  onChange: () => void;
  onFocus: () => void;
  onBlur: () => void;
}> = ({ onChange, onFocus, onBlur }) => (
  <div>
    <Box mb={3}>
      <Input placeholder="Input with onChange event (see Action Logger)" onChange={onChange} />
    </Box>
    <Box mb={3}>
      <Input
        placeholder="Input with onFocus & onBlur events (see Action Logger)"
        onFocus={onFocus}
        onBlur={onBlur}
      />
    </Box>
    <Box mb={3}>
      <InputWrapper />
    </Box>
  </div>
);
