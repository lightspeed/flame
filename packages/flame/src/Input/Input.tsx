import * as React from 'react';
import { css } from '@emotion/core';
import styled from '@emotion/styled';

import { themeGet } from '@styled-system/theme-get';
import { layout, LayoutProps, zIndex, ZIndexProps, compose } from 'styled-system';
import { Merge } from 'type-fest';

import { Flex, Box, border, BorderProps } from '../Core';
import { Label, FormHelper } from '../FormField';

import { IconVerified } from '../Icon/Verified';
import { IconWarning } from '../Icon/Warning';
import { IconDanger } from '../Icon/Danger';

type StatusType = 'valid' | 'error' | 'warning';

type StyledInputProps = {
  /** Set the size of the input */
  inputSize?: 'small' | 'regular' | 'large';
  hasSuffix?: boolean;
  hasPrefix?: boolean;
  /** Sets the element to read-only. The content will be selectable, but not editable */
  readOnly?: boolean;
};
const StyledInput = styled('input')<StyledInputProps>`
  width: 100%;
  flex: 1 1 0%;
  box-sizing: border-box;
  height: ${themeGet('space.6')};
  font-family: ${themeGet('fontFamily.sans-serif')};
  font-size: ${themeGet('fontSizes.text-s')};
  padding-left: ${themeGet('space.2')};
  padding-right: ${themeGet('space.2')};
  border: solid 1px transparent;
  background: transparent;
  color: ${themeGet('inputStyles.color')};
  min-width: 0;
  transition: all ${themeGet('transition.transition-duration-fast')} ease-in-out;

  &:disabled {
    opacity: 1;
  }

  ${props => {
    switch (props.inputSize) {
      case 'small':
        return css`
          font-size: ${themeGet('fontSizes.text-xs')(props)};
          height: ${themeGet('space.5')(props)};
        `;
      case 'large':
        return css`
          height: ${themeGet('space.7')(props)};
        `;
      case 'regular':
      default:
        return css``;
    }
  }}

  ${props =>
    props.hasPrefix &&
    css`
      padding-left: 0;
    `};

  ${props =>
    props.hasSuffix &&
    css`
      padding-right: 0;
    `};

  &::placeholder {
    color: ${themeGet('inputStyles.placeholder.color')};
  }

  &:focus,
  &:active {
    outline: none;
  }

  ${props =>
    props.readOnly &&
    css`
      color: ${themeGet('inputStyles.readonly.color')(props)};
    `};

  &:not([disabled]):not([readonly]):hover + div {
    border-color: ${themeGet('inputStyles.hover.border')};
  }

  /* Prettier does some nasty things if we merge both selectors... */
  &:not([disabled]):not([readonly]):focus + div {
    border-color: ${themeGet('inputStyles.focus.border')};
  }

  &:not([disabled]):not([readonly]):active + div {
    border-color: ${themeGet('inputStyles.active.border')};
  }
`;

interface InputBackdropProps extends BorderProps {
  status?: StatusType;
}
export const InputBackdrop = styled('div')<InputBackdropProps>`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  pointer-events: none;
  border-radius: ${themeGet('radii.radius-1')};
  background: transparent;
  box-shadow: ${themeGet('inputStyles.boxShadow')};
  border: solid 1px ${themeGet('inputStyles.border')};
  transition: border-color ${themeGet('transition.transition-duration-fast')} ease-in-out;

  ${props =>
    props.status === 'error' &&
    css`
      border-color: ${themeGet('inputStyles.error.border')(props)} !important;
    `};
  ${border}
`;

interface WrapperProps extends BorderProps, LayoutProps, ZIndexProps {
  disabled: boolean;
  readOnly: boolean;
  isAutofilled: boolean;
  status?: StatusType;
}
const Wrapper = styled('div')<WrapperProps>`
  position: relative;
  display: flex;
  align-items: center;
  border-radius: ${themeGet('radii.radius-2')};
  background: ${themeGet('inputStyles.background')};
  transition: all ${themeGet('transition.transition-duration-fast')} ease-in-out;
  width: 100%;

  ${compose(
    layout,
    border,
    zIndex,
  )}

  ${props =>
    props.disabled &&
    css`
      color: ${themeGet('inputStyles.disabled.color')(props)};
      background: ${themeGet('inputStyles.disabled.background')(props)};
      -webkit-text-fill-color: currentColor;
    `};

  ${props =>
    props.readOnly &&
    css`
      background: ${themeGet('inputStyles.readonly.background')(props)};
    `};

  ${props =>
    props.isAutofilled &&
    css`
      background: ${themeGet('inputStyles.autofilled.background')(props)};
      background-image: none;
      color: ${themeGet('inputStyles.autofilled.color')(props)};
    `};

  ${props => {
    switch (props.status) {
      case 'error':
        return css`
          .cr-input__status-icon {
            fill: ${themeGet('inputStyles.error.border')(props)};
          }
        `;
      case 'valid':
        return css`
          .cr-input__status-icon {
            fill: ${themeGet('inputStyles.valid.color')(props)};
          }
        `;
      case 'warning':
        return css`
          .cr-input__status-icon {
            fill: ${themeGet('inputStyles.warning.color')(props)};
          }
        `;
      default:
        return '';
    }
  }};
`;

interface StatusIconProps {
  status?: StatusType;
}
const StatusIcon: React.FC<StatusIconProps> = ({ status }) => {
  switch (status) {
    case 'valid':
      return <IconVerified data-testid="icon-valid" className="cr-input__status-icon" />;
    case 'error':
      return <IconDanger data-testid="icon-error" className="cr-input__status-icon" />;
    case 'warning':
      return <IconWarning data-testid="icon-warning" className="cr-input__status-icon" />;
    default:
      return null;
  }
};

export type BaseInputProps = Merge<
  Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'prefix'>,
  Omit<StyledInputProps, 'inputSize'>
> &
  InputBackdropProps & {
    size?: 'small' | 'regular' | 'large';
    status?: StatusType;
    disabled?: boolean;
    /** (Legacy) Was this element automatically filled by the browser? */
    isAutofilled?: boolean;
    /** Element to show inside the input on the left hand side. Usually reserved for Icons */
    prefix?: React.ReactNode;
    /** Element to show inside the input on the right hand side. Usually reserved for Icons */
    suffix?: React.ReactNode;
    css?: any;
  };
export const BaseInput = React.forwardRef<HTMLInputElement, BaseInputProps>(
  (
    {
      size,
      status,
      disabled,
      readOnly,
      isAutofilled,
      prefix,
      suffix,
      borderLeft,
      borderRight,
      borderRadius,
      borderTopLeftRadius,
      borderTopRightRadius,
      borderBottomLeftRadius,
      borderBottomRightRadius,
      width,
      ...restProps
    },
    ref,
  ) => {
    return (
      <Wrapper
        disabled={disabled}
        readOnly={readOnly}
        isAutofilled={isAutofilled}
        status={status}
        width={width}
        borderLeft={borderLeft}
        borderRight={borderRight}
        borderRadius={borderRadius}
        borderTopLeftRadius={borderTopLeftRadius}
        borderTopRightRadius={borderTopRightRadius}
        borderBottomLeftRadius={borderBottomLeftRadius}
        borderBottomRightRadius={borderBottomRightRadius}
      >
        {prefix && (
          <Flex flex={0} pl={2} pr={1} fontSize="text-s">
            {prefix}
          </Flex>
        )}
        <StyledInput
          ref={ref}
          disabled={disabled}
          inputSize={size}
          hasPrefix={!!prefix}
          hasSuffix={!!status || !!suffix}
          readOnly={readOnly}
          {...restProps}
        />
        <InputBackdrop
          status={status}
          borderLeft={borderLeft}
          borderRight={borderRight}
          borderRadius={borderRadius}
          borderTopLeftRadius={borderTopLeftRadius}
          borderTopRightRadius={borderTopRightRadius}
          borderBottomLeftRadius={borderBottomLeftRadius}
          borderBottomRightRadius={borderBottomRightRadius}
        />
        {(status || suffix) && (
          <Flex flex={0} pr={2} fontSize="text-s">
            {suffix}
            {status && (
              <Box ml={2}>
                <StatusIcon status={status} />
              </Box>
            )}
          </Flex>
        )}
      </Wrapper>
    );
  },
);

export interface InputProps extends Omit<BaseInputProps, 'status' | 'hasPrefix' | 'hasSuffix'> {
  /** The text label that will appear above the Input */
  label?: React.ReactNode;
  /** Helper element that will be positioned to the top right of the Input */
  labelHelper?: React.ReactNode;
  /** Paragraph that will appear beneath the label */
  description?: React.ReactNode;
  /** If a status is present, this content will appear on the bottom of the Input */
  statusMessage?: React.ReactNode;
  /** Paragraph that will appear on the bottom of the Input */
  textHelper?: React.ReactNode;
  /** Set the status of the Input */
  status?:
    | StatusType
    | {
        type?: StatusType;
        message?: React.ReactNode;
      };
  css?: any;
}

/**
 * The Acceptor of User Information, Messenger of the Internet Gods.
 */
export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (
    { label, id, status, description, statusMessage, textHelper, labelHelper, ...restProps },
    ref,
  ) => {
    const labelId = id && label ? `${id}-label` : undefined;
    const descriptionId = id && description ? `${id}-description` : undefined;
    const nextStatus = status && typeof status === 'object' ? status.type : (status as StatusType);
    const nextMessage = status && typeof status === 'object' ? status.message : statusMessage;

    return (
      <React.Fragment>
        {label && (
          <Label
            id={labelId}
            htmlFor={id}
            description={description}
            descriptionProps={{ id: descriptionId }}
          >
            {labelHelper ? (
              <Flex
                width="100%"
                flexDirection="row"
                justifyContent="space-between"
                alignItems="center"
              >
                <div>{label}</div>
                <div>{labelHelper}</div>
              </Flex>
            ) : (
              label
            )}
          </Label>
        )}
        <BaseInput
          ref={ref}
          id={id}
          status={nextStatus}
          aria-labelledby={labelId}
          aria-describedby={descriptionId}
          {...restProps}
        />
        {status && nextMessage && (
          <FormHelper mt={1} status={nextStatus}>
            {nextMessage}
          </FormHelper>
        )}
        {textHelper && !nextMessage && <FormHelper mt={1}>{textHelper}</FormHelper>}
      </React.Fragment>
    );
  },
);

// Keeping flameName for compat with Group & inputBlock prop. Will be deprecated.
// @ts-ignore
Input.flameName = 'Input';
