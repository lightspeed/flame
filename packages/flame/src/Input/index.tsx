import { BaseInput, BaseInputProps, Input, InputProps } from './Input';

export { BaseInput, BaseInputProps, Input, InputProps };
