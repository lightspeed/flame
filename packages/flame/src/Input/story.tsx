import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { action, decorateAction } from '@storybook/addon-actions';
import { withReadme } from 'storybook-readme';

import { Input } from './Input';
import Readme from './README.md';
import { Box } from '../Core';
import { Badge } from '../Badge';
import { Button } from '../Button';
import { Icon } from '../Icon';

const stories = storiesOf('Components|Input', module).addDecorator(withReadme(Readme));

const firstArgAction = decorateAction([(args: any) => [args[0].target.value]]);

type State = {
  status: {
    type?: 'error' | 'valid' | 'warning';
    message?: string;
  };
  input: any;
};
class InputWrapper extends React.Component<any, State> {
  constructor(props: any) {
    super(props);
    this.state = {
      status: {},
      input: '',
    };
  }

  handleClick() {
    if (!this.state.input || Number.isNaN(parseFloat(this.state.input))) {
      this.setState({ status: { type: 'error', message: 'Please enter a valid number' } });
    } else if (this.state.input < 10) {
      this.setState({ status: { type: 'warning', message: 'You can do more than 10!' } });
    } else {
      this.setState({ status: { type: 'valid', message: 'Good job!' } });
    }
  }

  handleChange(e: React.FormEvent<HTMLInputElement>) {
    const target = e.target as HTMLInputElement;
    this.setState({ input: target.value });
  }

  handleFocus() {
    this.setState({ status: {} });
  }

  render() {
    return (
      <div>
        <Input
          placeholder="Input with error message..."
          id="error"
          onChange={e => this.handleChange(e)}
          onFocus={() => this.handleFocus()}
          label="Enter a number"
          status={this.state.status.type as any}
          statusMessage={this.state.status.message}
        />
        <br />
        <Button onClick={() => this.handleClick()}>Submit</Button>
      </div>
    );
  }
}

stories.add('Story', () => (
  <div>
    <Box mb={3}>
      <h3>Simple text input</h3>
      <Input id="simple" label="Label" placeholder="Placeholder text..." />
    </Box>
    <Box mb={3}>
      <h3>Advanced text input</h3>
      <Input
        id="advanced"
        label="Label"
        labelHelper={
          <Badge type="info" size="small">
            Label Helper
          </Badge>
        }
        textHelper="Text Helper"
        description="Description text"
        placeholder="Placeholder text..."
        prefix="$"
        suffix="*"
      />
    </Box>

    <h2>States</h2>
    <Box mb={3}>
      <Input placeholder="Placeholder text..." />
    </Box>
    <Box mb={3}>
      <Input placeholder="Hover input..." />
    </Box>
    <Box mb={3}>
      <Input placeholder="Focus input..." />
    </Box>
    <Box mb={3}>
      <Input readOnly value="Readonly input" />
    </Box>
    <Box mb={3}>
      <Input disabled placeholder="Disabled input..." />
    </Box>
    <Box mb={3}>
      <Input
        status="valid"
        placeholder="Valid input..."
        prefix={<Icon name="customers" />}
        suffix={<Icon name="info" color="secondary" />}
      />
    </Box>
    <Box mb={3}>
      <Input status="warning" placeholder="Warning input..." />
    </Box>
    <Box mb={3}>
      <Input status="error" placeholder="Error input..." />
    </Box>

    <div>
      <h2>Size</h2>
      <Box mb={3}>
        <Input size="small" placeholder="Small input..." />
      </Box>
      <Box mb={3}>
        <Input placeholder="Normal input..." />
      </Box>
      <Box mb={3}>
        <Input size="large" placeholder="Large input..." />
      </Box>
    </div>
    <div>
      <h2>Input Error Text</h2>
      <Box mb={3}>
        <Input
          placeholder="This is some text on an input that has an error"
          status="error"
          statusMessage="This is an error message"
        />
      </Box>
    </div>
    <div>
      <h2>Input Error Text (using the legacy API)</h2>
      <Box mb={3}>
        <Input
          placeholder="This is some text on an input that has an error"
          status={{ type: 'error', message: 'This is an error message' }}
        />
      </Box>
    </div>
  </div>
));

stories.add(
  'Events',
  () => (
    <div>
      <Box mb={3}>
        <Input
          placeholder="Input with onChange event (see Action Logger)"
          onChange={firstArgAction('onChange')}
        />
      </Box>
      <Box mb={3}>
        <Input
          placeholder="Input with onFocus & onBlur events (see Action Logger)"
          onFocus={action('onFocus')}
          onBlur={action('onBlur')}
        />
      </Box>
      <Box mb={3}>
        <InputWrapper />
      </Box>
    </div>
  ),
  { percy: { skip: true } },
);
