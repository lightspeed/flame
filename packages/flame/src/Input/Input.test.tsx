import * as React from 'react';
import { createComponent, customRender, fireEvent } from 'test-utils';
import { Input } from './Input';

type BaseProps = {
  id: string;
  name: string;
  placeholder: string;
};

describe('<Input />', () => {
  const baseProps: BaseProps = {
    id: 'username',
    name: 'username',
    placeholder: 'Placeholder text',
  };

  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const component = createComponent(<Input {...baseProps} />);
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  it('should have rendered an input of type text', () => {
    const { getByPlaceholderText } = customRender(
      <Input {...baseProps} placeholder="test-input" size="large" />,
    );
    const element: any = getByPlaceholderText('test-input');
    expect(element.type).toBe('text');
  });

  it('should not crash on null status', () => {
    const { getByPlaceholderText } = customRender(
      <Input {...baseProps} status={null} placeholder="test-input" size="large" />,
    );
    const element: any = getByPlaceholderText('test-input');
    expect(element.type).toBe('text');
  });

  it('should forward the disabled property to the base input', () => {
    const { getByPlaceholderText } = customRender(
      <Input {...baseProps} disabled placeholder="test-input" size="small" />,
    );
    const element: any = getByPlaceholderText('test-input');
    expect(element.disabled).toBeTruthy();
  });

  it('should forward the read-only property to the base input', () => {
    const { getByPlaceholderText } = customRender(
      <Input {...baseProps} readOnly placeholder="test-input" />,
    );

    expect(getByPlaceholderText('test-input')).toHaveAttribute('readonly');
  });

  it('should contain a prefix and a Suffix', () => {
    const { getByText } = customRender(
      <Input {...baseProps} prefix={<span>My Prefix</span>} suffix={<span>My Suffix</span>} />,
    );
    getByText('My Prefix');
    getByText('My Suffix');
  });

  describe('Input types and their messages', () => {
    it('should render a valid state', () => {
      const { getByText, getByTestId } = customRender(
        <Input {...baseProps} status="valid" statusMessage="Valid input." />,
      );
      getByText('Valid input.');
      getByTestId('icon-valid');
    });

    it('should render an errored state', () => {
      const { getByText, getByTestId } = customRender(
        <Input {...baseProps} status="error" statusMessage="Error input!" />,
      );
      getByText('Error input!');
      getByTestId('icon-error');
    });

    it('should render an warning state', () => {
      const { getByText, getByTestId } = customRender(
        <Input {...baseProps} status="warning" statusMessage="Warning input." />,
      );
      getByText('Warning input.');
      getByTestId('icon-warning');
    });

    it('should support the legacy API', () => {
      const { getByText, getByTestId } = customRender(
        <Input {...baseProps} status={{ type: 'warning', message: 'Warning input.' }} />,
      );
      getByText('Warning input.');
      getByTestId('icon-warning');
    });
  });

  describe('when passing ref to the component', () => {
    it('should pass the node correctly', () => {
      const ref = jest.fn();
      const { container } = customRender(<Input ref={ref} />);
      const input = container.querySelector('input');
      const [firstArg] = ref.mock.calls[0];

      expect(ref).toHaveBeenCalledTimes(1);
      expect(firstArg).toBe(input);
    });
  });

  describe('label', () => {
    it('Can find input by label', () => {
      const { getByText } = customRender(<Input label="My Label" />);

      expect(getByText('My Label')).toBeTruthy();
    });

    it('Should set htmlFor properly', () => {
      const { container, queryByLabelText } = customRender(<Input id="my-id" label="My Label" />);
      const input = queryByLabelText('My Label');
      const linkedLabel = container.querySelector(`label[for="${input.id}"]`);

      expect(input).toBeInTheDocument();
      expect(linkedLabel).toBeInTheDocument();
    });

    it('Should render aria-labelledBy properly', () => {
      const { container, queryByLabelText } = customRender(<Input id="my-id" label="My Label" />);
      const input = queryByLabelText('My Label');
      const linkedAriaText = container.querySelector(`#${input.getAttribute('aria-labelledBy')}`);

      expect(input).toBeInTheDocument();
      expect(linkedAriaText).toBeInTheDocument();
    });
  });

  describe('Events', () => {
    it('handles the `onChange` event', () => {
      const onChange = jest.fn();
      const { container } = customRender(<Input onChange={onChange} />);
      const input = container.querySelector('input');

      expect(onChange).not.toHaveBeenCalled();
      fireEvent.change(input, { target: { value: 'Awesome value' } });
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('handles the `onFocus` event', () => {
      const onFocus = jest.fn();
      const { container } = customRender(<Input onFocus={onFocus} />);
      const input = container.querySelector('input');

      expect(onFocus).not.toHaveBeenCalled();
      fireEvent.focus(input);
      expect(onFocus).toHaveBeenCalledTimes(1);
    });

    it('handles the `onBlur` event', () => {
      const onBlur = jest.fn();
      const { container } = customRender(<Input onBlur={onBlur} />);
      const input = container.querySelector('input');

      expect(onBlur).not.toHaveBeenCalled();
      fireEvent.blur(input);
      expect(onBlur).toHaveBeenCalledTimes(1);
    });
  });
});
