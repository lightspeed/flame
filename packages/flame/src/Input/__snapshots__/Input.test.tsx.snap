// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Snapshot should match snapshot 1`] = `
.emotion-2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all 100ms ease-in-out;
  transition: all 100ms ease-in-out;
  background: #fff;
  border-radius: 0.375rem;
  width: 100%;
}

.emotion-0 {
  color: #494c4c;
  width: 100%;
  -webkit-flex: 1 1 0%;
  -ms-flex: 1 1 0%;
  flex: 1 1 0%;
  box-sizing: border-box;
  border: solid 1px transparent;
  background: transparent;
  font-family: Lato,Helvetica Neue,Helvetica,Arial,sans-serif;
  min-width: 0;
  font-size: 1rem;
  height: 40px;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  -webkit-transition: all 100ms ease-in-out;
  transition: all 100ms ease-in-out;
}

@media screen and (min-width:601px) {
  .emotion-0 {
    font-size: 0.875rem;
    height: 36px;
  }
}

.emotion-0::-webkit-input-placeholder {
  color: #abb3b3;
}

.emotion-0::-moz-placeholder {
  color: #abb3b3;
}

.emotion-0:-ms-input-placeholder {
  color: #abb3b3;
}

.emotion-0::placeholder {
  color: #abb3b3;
}

.emotion-0:disabled {
  opacity: 1;
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:active {
  outline: none;
}

.emotion-0:not([disabled]):not([readonly]):hover + div {
  border-color: #9ba0ab;
}

.emotion-0:not([disabled]):not([readonly]):focus + div {
  border-color: #2875c6;
}

.emotion-0:not([disabled]):not([readonly]):active + div {
  border-color: #2875c6;
}

.emotion-1 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  pointer-events: none;
  border-radius: 0.1875rem;
  background: transparent;
  box-shadow: inset 0 0.0625rem 0.125rem rgba(56,61,61,0.1);
  border: solid 1px #c4cacc;
  -webkit-transition: border-color 100ms ease-in-out;
  transition: border-color 100ms ease-in-out;
}

<div
  className="emotion-2"
>
  <input
    className="emotion-0"
    id="username"
    name="username"
    placeholder="Placeholder text"
  />
  <div
    className="emotion-1"
  />
</div>
`;
